vitessio/arewefastyet
# Fix: command microbench run not returning status 0 when pkg path is incorrect

## Description

This pull request fixes the bug described in #88. The bug was that no output nor error code was produced by the CLI's `microbench run` command when feeding it a wrong path for the packages. In short, the following command behaved like everything went well:

```sh
arewefastyetcli microbench run invalid_path output.txt --config /tmp/config.yaml
```

## How to fix it?

In the `./go/tools/microbench` package's `MicroBenchmark` function, we load the packages we want to use. Loading packages with `packages.Load()` does not necessarily return a non-nil error if the given packages are incorrect.

```go
// below, the package.Load function's signature

func Load(cfg *Config, patterns ...string) ([]*Package, error)
```

However, the array of packages that we get (`[]*packages.Package`) will be populated with the packages we gave in the `patterns` argument of the function `packages.Load()`. The elements of the array contain a field `Errors` (an array of `package.Error`).

I then decided to verify this field for each package we treat. If the field contains one or more error, I append it to a slice of `error`, which eventually gets transform into a single `error` that we will return to the caller of function `MicroBenchmark`.

## Deal with errors

Another side of this issue was that errors were not handled in the `MicroBenchmark` function, which would instead `panic` or `log.Fatal`. This behavior was removed, instead, we return an error.

This allows us to treat error even higher in the stack, "microbench run" Cobra's command functions are now dealing with the error.

## Concatenation of errors

As mentioned earlier, in order to warn the user of all the errors we encountered with the packages and have a higher level of verbose, all the errors from `package.Package` are appended to a slice, which is then transformed into a single `error`. This is achieved by a newly added package in `./go/tools/errors`. This package will contain the functions and methods we need to ease the work we do around errors.


## How to reproduce and verify the fix

A new unit test was added for our `MicroBenchmark` function, the test contains a single test case and verifies that the function behaves correctly when given a wrong package path.

```sh
go test -v ./go/... -run ^TestMicroBenchmark$
```

Additionally, the fix can be verified by running:

```sh
go run ./go/main.go microbench run INVALID output.txt --config config/config.yaml 2>&1 >/dev/null | head -n 2
```


Which should output the following content:

```
Error: invalid package parsing:
package INVALID is not in GOROOT (/usr/local/go/src/INVALID)
```

## Related issues

Fixes #88.
